src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,208 @@
+                     pswfcs, pswfc_ang_moms, pswfc_occs, rvlocpzdr, rprojsdr, identifier)
+end
+
+function parse_upf_file(path; identifier=path)
+    pseudo = load_upf(path)
+    # Pseudo-ion charge
+    Zion = Int(pseudo["header"]["z_valence"])
+    # Radial grid
+    rgrid = pseudo["radial_grid"]
+    drgrid = pseudo["radial_grid_derivative"]
+    # Maximum angular momentum channel
+    lmax = pseudo["header"]["l_max"]
+    # Number of Kleinman-Bylander β projectors
+    nproj = pseudo["header"]["number_of_proj"]
+    # Number of pseudo-atomic wavefunctions
+    nwfc = pseudo["header"]["number_of_wfc"]
+    # Local potential
+    vloc = pseudo["local_potential"]
+
+    # Kleinman-Bylander β projectors
+    projs = [Vector[] for _ = 0:lmax]
+    for i = 1:nproj
+        proj_data = pseudo["beta_projectors"][i]
+        l = proj_data["angular_momentum"]
+        proj = proj_data["radial_function"]
+        push!(projs[l+1], proj)
+    end
+    
+    # β-projector coupling coefficients
+    h = Matrix[]
+    n = 1
+    for l = 0:lmax
+        nproj_l = length(projs[l+1])
+        push!(h, pseudo["D_ion"][n:n+nproj_l-1, n:n+nproj_l-1] * 2)
+        n += nproj_l
+    end
+    
+    # Pseudo-atomic wavefunctions
+    pswfcs = Vector{Vector}(undef, nwfc)
+    pswfc_ang_moms = Vector(undef, nwfc)
+    pswfc_occs = Vector(undef, nwfc)
+    for i = 1:nwfc
+        pswfc_data = pseudo["atomic_wave_functions"][i]
+        pswfcs[i] = pswfc_data["radial_function"]
+        pswfc_ang_moms[i] = pswfc_data["angular_momentum"]
+        pswfc_occs[i] = pswfc_data["occupation"]
+    end
+
+    # Useful precomputed quantities
+    rvlocpzdr = (rgrid .* vloc .+ 2Zion) .* drgrid
+    rprojsdr = [Vector[] for _ = 0:lmax]
+    for l = 0:lmax
+        for proj in projs[l+1]
+            push!(rprojsdr[l+1], proj .* view(rgrid, 1:length(proj)) .* view(drgrid, 1:length(proj)))
+        end
+    end
+
+    PspUpf(Zion, lmax, rgrid, drgrid, vloc, projs, h,
+           pswfcs, pswfc_ang_moms, pswfc_occs, rvlocpzdr, rprojsdr; identifier)
+end
+
+charge_ionic(psp::PspUpf) = psp.Zion
+
+"""
+    eval_psp_projector_real(psp::PspUpf, i::Number, l::Number, r::Number)
+
+Evaluate the ith Kleinman-Bylander β projector with angular momentum l at real-space distance r
+via cubic spline interpolation on the real-space mesh.
+
+!!! Note
+UPFs store `rgrid[i] * β[l,n](rgrid[i])`, so we must divide by `r`.
+"""
+function eval_psp_projector_real(psp::PspUpf, i, l, r::T) where {T <: Real}
+    ir_cut = length(psp.projs[l+1][i])
+    linear_interpolation((psp.rgrid[1:ir_cut],), psp.projs[l+1][i])(r) / r
+end
+
+"""
+    eval_psp_projector_fourier(psp::PspUPF, i::Number, l::Number, q::Number)
+Evaluate the ith Kleinman-Bylander β projector with angular momentum l in at k-space distance q.
+
+!!! Note
+UPFs store `rgrid[i] * β[l,n](rgrid[i])`, so the integrand has `r` instead of `r^2`
+"""
+function eval_psp_projector_fourier(psp::PspUpf, i, l, q::T)::T where {T <: Real}
```

Comment:
I think I would introduce the Val already at this level and drop the `psp_projector_fouier_inner_loop` function, e.g. just have

```julia
function eval_psp_projector_fourier(psp::PspUpf, i, l, q::T)::T where {T <: Real}
    eval_psp_projector_fourier(psp, i, Val(l), q)
end

function eval_psp_projector_fourier(psp::PspUpf, i, order::Val{l}, q::T)::T where {T <: Real, l}
    ...
end
```

Reply:
One minor thing we might consider doing is to evaluate all projectors of the same AM `l` at once. That avoids having to recompute the bessel function for each projector `i`. Given that `eval_psp_projector_fourier` is only called in a loop over `i` it is also not too hard to implement that. OTOH the potential savings is small (as there is usually just 1-3 `i`s).